### PR TITLE
Add New Book

### PR DESCRIPTION
**Pull request by gastonmaspero**
## Summary ##

**Trello card**: https://trello.com/c/EQdIcjqR/25-n-8-vista-para-sugerir-un-nuevo-libro

Description: This feature adds the "Add new book" view, and all its logic using the reactive paradigm.

### iPhone X

Simulator Screen Shot - iPhone X - 2019-06-11 at 21 04 14
screencast

screencast2

### CODE REVIEW

File: WBooks/AddNew/AddNewController.swift [hunk 1]
```diff
@@ -7,23 +7,165 @@
...
+import ReactiveSwift
+import ReactiveCocoa
+import Result
+import WolmoCore
 
 class AddNewController: UIViewController {
-
+    
+    private let formFilled = MutableProperty(false)
+
```
Comment (carcos-wolox) [inline WBooks/AddNew/AddNewController.swift]: Try to avoid this lines

File: WBooks/AddNew/AddNewController.swift [hunk 2]
```diff
@@ -7,23 +7,165 @@
...
 import UIKit
+import ReactiveSwift
+import ReactiveCocoa
+import Result
+import WolmoCore
 
 class AddNewController: UIViewController {
-
+    
+    private let formFilled = MutableProperty(false)
```
Comment (carcos-wolox) [inline WBooks/AddNew/AddNewController.swift]: private properties should have _ at the beginning

File: WBooks/AddNew/AddNewController.swift [hunk 3]
```diff
@@ -7,23 +7,165 @@
...
+    
+    private var loading = false
```
Comment (carcos-wolox) [inline WBooks/AddNew/AddNewController.swift]: same

File: WBooks/AddNew/AddNewController.swift [hunk 4]
```diff
@@ -7,23 +7,165 @@
...
+        _view.setView()
+        setBindings()
+        setFormValidation()
+        setSubmitButton()
+        setImagePicker()
+    }
+    
+    override func viewWillAppear(_ animated: Bool) {
+        super.viewWillAppear(animated)
+        configureNavBar()
```
Comment (carcos-wolox) [inline WBooks/AddNew/AddNewController.swift]: Why you need to do this each time the view appears and not in viewDidLoad?

File: WBooks/AddNew/AddNewController.swift [hunk 5]
```diff
@@ -7,23 +7,165 @@
...
+        _viewModel.bookName <~ _view.booksName.reactive.continuousTextValues.skipNil()
+        _viewModel.author <~ _view.author.reactive.continuousTextValues.skipNil()
+        _viewModel.year <~ _view.year.reactive.continuousTextValues.skipNil()
+        _viewModel.genre <~ _view.genre.reactive.continuousTextValues.skipNil()
+        _viewModel.descrpition <~ _view.bookDescription.reactive.continuousTextValues.skipNil()
+    }
+    
+    private func setFormValidation() {
+        let bookNameValidation = _view.booksName.reactive.continuousTextValues.skipNil().map { text in
+            !text.isEmpty
```
Comment (carcos-wolox) [inline WBooks/AddNew/AddNewController.swift]: I think wolmo core has .isNotEmpty for strings

File: WBooks/AddNew/AddNewController.swift [hunk 6]
```diff
@@ -7,23 +7,165 @@
...
+            if let this = self {
+                if formFilled && !this.loading {
+                    this._view.enableSubmit()
+                } else {
+                    this._view.disableSubmit()
+                }
+            }
+        }
+        _view.submitButton.reactive.controlEvents(.touchUpInside).observeValues { [weak self] _ in
+            self?.loading = true
```
Comment (carcos-wolox) [inline WBooks/AddNew/AddNewController.swift]: You could handle this with reactive swift in a simpler way. If you want, you can search it, but it's fine if you leave this in training.

File: WBooks/AddNew/AddNewController.swift [hunk 7]
```diff
@@ -7,23 +7,165 @@
...
+    func onAddNewSuccess() {
+        loading = false
+        _view.resetForm()
+        _view.enableInteractions()
+        let alert = UIAlertController(alertViewModel: ErrorAlertViewModel(title: "THANKS".localized(), message: "BOOK_ADDED".localized(), dismissButtonTitle: "ACCEPT".localized()))
+        self.present(alert, animated: true, completion: nil)
+    }
+    
+    func onError(error: Error) {
+        loading = true
```
Comment (carcos-wolox) [inline WBooks/AddNew/AddNewController.swift]: I think that if you got an error you wont be able to submit again if this is true 🤔

File: WBooks/AddNew/AddNewController.swift [hunk 8]
```diff
@@ -7,23 +7,165 @@
...
+        tapRecognizer.delegate = self
+        _view.cover.addGestureRecognizer(tapRecognizer)
+        _view.cover.isUserInteractionEnabled = true
+    }
+    
+    @objc private func imageTapped(sender: AnyObject) {
+        let alertController = UIAlertController(title: .none, message: .none, preferredStyle: .actionSheet)
+        
+        //Gallery option
+        let chooseAction = UIAlertAction(title: "Gallery", style: .default) { [weak self] _ in
```
Comment (carcos-wolox) [inline WBooks/AddNew/AddNewController.swift]: Add string to localized

File: WBooks/AddNew/AddNewController.swift [hunk 9]
```diff
@@ -7,23 +7,165 @@
...
+            let imagePickerController = UIImagePickerController.imagePicker
+            imagePickerController.delegate = self
+            imagePickerController.sourceType = .photoLibrary
+            self?.present(imagePickerController, animated: true, completion: .none)
+        }
+        //Camera option
+        if UIImagePickerController.isSourceTypeAvailable(.camera) {
+            let takeAction = UIAlertAction(title: "Camera", style: .default) { [weak self] _ in
```
Comment (carcos-wolox) [inline WBooks/AddNew/AddNewController.swift]: same

File: WBooks/AddNew/AddNewController.swift [hunk 10]
```diff
@@ -7,23 +7,165 @@
...
+                let imagePickerController = UIImagePickerController.imagePicker
+                imagePickerController.delegate = self
+                imagePickerController.sourceType = .camera
+                self?.present(imagePickerController, animated: true, completion: .none)
+            }
+            alertController.addAction(takeAction)
+        }
+        //Cancel option
+        let cancelAction = UIAlertAction(title: "Cancel", style: .cancel, handler: .none)
```
Comment (carcos-wolox) [inline WBooks/AddNew/AddNewController.swift]: same

File: WBooks/AddNew/AddNewController.swift [hunk 11]
```diff
@@ -7,23 +7,165 @@
...
+            let imagePickerController = UIImagePickerController.imagePicker
+            imagePickerController.delegate = self
+            imagePickerController.sourceType = .photoLibrary
+            self?.present(imagePickerController, animated: true, completion: .none)
+        }
+        //Camera option
+        if UIImagePickerController.isSourceTypeAvailable(.camera) {
+            let takeAction = UIAlertAction(title: "Camera", style: .default) { [weak self] _ in
+                let imagePickerController = UIImagePickerController.imagePicker
+                imagePickerController.delegate = self
```
Comment (carcos-wolox) [inline WBooks/AddNew/AddNewController.swift:132]: This is not required

File: WBooks/AddNew/AddNewController.swift [hunk 12]
```diff
@@ -7,23 +7,165 @@
...
+        _view.cover.image = info[UIImagePickerControllerOriginalImage] as? UIImage
+        dismiss(animated: true, completion: nil)
     }
-    */
+}
+
+extension AddNewController: UINavigationControllerDelegate {
+}
 
+extension AddNewController: UIGestureRecognizerDelegate {
```
Comment (carcos-wolox) [inline WBooks/AddNew/AddNewController.swift]: This is not required

File: WBooks/AddNew/AddNewController.swift [hunk 13]
```diff
@@ -7,23 +7,165 @@
...
-        // Pass the selected object to the new view controller.
+extension AddNewController: UIImagePickerControllerDelegate {
+    func imagePickerController(_ picker: UIImagePickerController, didFinishPickingMediaWithInfo info: [String: Any]) {
+        _view.cover.image = info[UIImagePickerControllerOriginalImage] as? UIImage
+        dismiss(animated: true, completion: nil)
     }
-    */
+}
+
+extension AddNewController: UINavigationControllerDelegate {
```
Comment (carcos-wolox) [inline WBooks/AddNew/AddNewController.swift]: Why is this required?

File: WBooks/AddNew/AddNewViewModel.swift [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+class AddNewViewModel {
+    
+    let bookName = MutableProperty("")
+    let author = MutableProperty("")
+    let year = MutableProperty("")
+    let genre = MutableProperty("")
+    let descrpition = MutableProperty("")
+    
+    private var book: Book {
+        return Book(title: bookName.value, author: author.value, id: 0, genre: genre.value, year: year.value, image: "some url", status: "available")
```
Comment (carcos-wolox) [inline WBooks/AddNew/AddNewViewModel.swift]: Break this line

File: WBooks/Models/BookRepository.swift [hunk 15]
```diff
@@ -41,4 +41,28 @@ class BookRepository {
...
         }
     }
+    
+    static func addNewBook(book: Book, onSuccess: @escaping () -> Void, onError: @escaping (Error) -> Void) {
+        let url = URL(string: "https://swift-training-backend.herokuapp.com/books")!
+        
+        let parameters = [
+            "author": book.author,
+            "title": book.title,
+            "image": "some url",
```
Comment (carcos-wolox) [inline WBooks/Models/BookRepository.swift]: book already has this value

File: WBooks/Models/BookRepository.swift [hunk 16]
```diff
@@ -41,4 +41,28 @@ class BookRepository {
...
+            "year": book.year,
+            "genre": book.genre,
+            "status": "available"
```
Comment (carcos-wolox) [inline WBooks/Models/BookRepository.swift]: same

File: WBooks/Library/MainMenuController.swift [hunk 17]
```diff
@@ -48,6 +53,14 @@ class MainMenuController: UIViewController {
         _view.tableView.register(cell: BookCell.self)
         _view.tableView.backgroundColor = UIColor.clear
     }
+    
+    private func configureNavBar() {
+        self.tabBarController?.setNavigationBarTitle("NAVIGATION_BAR_TITLE_LIBRARY".localized(), font: UIFont.systemFont(ofSize: 18, weight: UIFont.Weight.medium), color: UIColor.white)
```
Comment (carcos-wolox) [inline WBooks/Library/MainMenuController.swift]: You are adding this for all View Controllers, you could add this function in an UIViewControllerExtension

**Review by carcos-wolox — CHANGES_REQUESTED** [review]
Nice! just some comments

File: WBooks/Text Field/TextFieldExtension.swift [hunk 18]
```diff
@@ -0,0 +1,21 @@
...
+
+import Foundation
+import UIKit
+
+extension UITextField {
+    func setBottomBorder(borderColor: UIColor, width: Double) {
+        borderStyle = UITextBorderStyle.none
+        backgroundColor = UIColor.clear
+        let borderLine = UIView()
+        borderLine.frame = CGRect(x: 0, y: Double(self.frame.height) - width, width: Double(self.frame.width), height: width)
```
Comment (carcos-wolox) [inline WBooks/Text Field/TextFieldExtension.swift]: unnecessary self

File: WBooks/Text Field/TextFieldExtension.swift [hunk 19]
```diff
@@ -0,0 +1,21 @@
...
+        borderLine.backgroundColor = borderColor
+        self.addSubview(borderLine)
```
Comment (carcos-wolox) [inline WBooks/Text Field/TextFieldExtension.swift]: same

File: WBooks/AddNew/AddNewController.swift (hunk 4 above)
Comment (gastonmaspero) [reply WBooks/AddNew/AddNewController.swift]: This is because I need to change the title and buttons of the nav bar for each tab bar view. That nav bar belongs (I understand it does) to the tab controller, so if I include that code on view did load it configures the nav bar only when the view is loaded to memory, and thus the nav bar items don't change in subsequent view changes from the tab bar.

File: WBooks/AddNew/AddNewController.swift (hunk 7 above)
Comment (gastonmaspero) [reply WBooks/AddNew/AddNewController.swift]: That is a typo right there. It is supposed to be false. I'll change it!

File: WBooks/AddNew/AddNewController.swift (hunk 13 above)
Comment (gastonmaspero) [reply WBooks/AddNew/AddNewController.swift]: A compile time error triggers if that is not included. When I make the AddNewController the image picker delegate, it requires AddNewController to conform to UIImagePickerControllerDelegate and UINavigationControllerDelegate. I put it on a separate extension to conform to the Wolox good practices you mentioned. But I can include it on line 156 with the extension that conforms to UIImagePickerControllerDelegate.

File: WBooks/AddNew/AddNewController.swift (hunk 4 above)
Comment (carcos-wolox) [reply WBooks/AddNew/AddNewController.swift]: I think each of the should be in a different navigation controller

File: WBooks/AddNew/AddNewController.swift (hunk 13 above)
Comment (carcos-wolox) [reply WBooks/AddNew/AddNewController.swift]: Yes, if If you are not adding any function, you could set this with UIImagePickerControllerDelegate

**Review by carcos-wolox — APPROVED** [review]
Good!!!
